dump/mapping/bufr_adpupa_prepbufr.py

```diff
@@ -0,0 +1,291 @@
+
+        # Loop through each variable and add it to the description
+        for var in variables:
+            description.add_variable(
+                name=var['name'],
+                source=var['source'],
+                units=var['units'],
+                longName=var['longName']
+            )
+
+    return description
+
+
+def _make_obs(comm, input_path, mapping_path, cycle_time):
+    """
+    Create the ioda adpupa prepbufr observations:
+    - reads values
+
+    Parameters
+    ----------
+    comm: object
+            The communicator object (e.g., MPI)
+    input_path: str
+            The input bufr file
+    mapping_path: str
+            The input bufr2ioda mapping file
+    cycle_time: str
+            The cycle in YYYYMMDDHH format
+    """
+
+    # Get container from mapping file first
+    logging(comm, 'INFO', 'Get container from bufr')
+    container = bufr.Parser(input_path, mapping_path).parse(comm)
+
+    logging(comm, 'DEBUG', f'container list (original): {container.list()}')
+    logging(comm, 'DEBUG', f'prepbufrDataLevelCategory')
+    cat = container.get('variables/prepbufrDataLevelCategory')
+ 
+    logging(comm, 'DEBUG', f'Change longitude range from [0,360] to [-180,180]')
+    lon = container.get('variables/longitude')
+    lon_paths = container.get_paths('variables/longitude')
+    lon[lon>180] -= 360
+
+    logging(comm, 'DEBUG', f'observationType')
+    typ = container.get('variables/observationType')
+    typ_paths = container.get_paths('variables/observationType')
+
+    logging(comm, 'DEBUG', f'Do ps calculation')
+    pob = container.get('variables/pressure')
+    pob_paths = container.get_paths('variables/pressure')
+    ps = np.full(pob.shape[0], pob.fill_value)
+    ps = np.where(cat == 0, pob, ps)
+
+    logging(comm, 'DEBUG', f'Do tsen and tv calculation')
+    tpc = container.get('variables/temperatureEventProgramCode')
+    tob = container.get('variables/airTemperature')
+    tob_paths = container.get_paths('variables/airTemperature')
+    tsen = np.full(tob.shape[0], tob.fill_value)
+    tsen = np.where(((tpc >= 1) & (tpc < 8)), tob, tsen)
+    tvo = np.full(tob.shape[0], tob.fill_value)
+    tvo = np.where((tpc == 8), tob, tvo)
+
+    logging(comm, 'DEBUG', f'Do ps QM calculations')
+    pqm = container.get('variables/pressureQualityMarker')
+    psqm = np.full(pqm.shape[0], pqm.fill_value)
+    psqm = np.where(cat == 0, pqm, psqm)
+
+    logging(comm, 'DEBUG', f'Do tsen and tv QM calculations')
+    tobqm = container.get('variables/airTemperatureQualityMarker')
+    tsenqm = np.full(tobqm.shape[0], tobqm.fill_value)
+    tsenqm = np.where(((tpc >= 1) & (tpc < 8)), tobqm, tsenqm)
+    tvoqm = np.full(tobqm.shape[0], tobqm.fill_value)
+    tvoqm = np.where((tpc == 8), tobqm, tvoqm)
+
+    logging(comm, 'DEBUG', f'Do ps ObsError calculations')
+    poe = container.get('variables/pressureError')
+    psoe = np.full(poe.shape[0], poe.fill_value)
+    psoe = np.where(cat == 0, poe, psoe)
```

Comment:
Is it possible to define these (and the corresponding container.replace) as a method?  There so many repeat codes here.